internal/controller/volumereplicationgroup_controller.go

```diff
@@ -800,41 +795,267 @@ func (v *VRGInstance) separatePVCsUsingVRGStatus(pvcList *corev1.PersistentVolum
+
+			replicationClassMatchFound := false
+
+			for _, replicationClass := range v.replClassList.Items {
+				if storageClass.Provisioner == replicationClass.Spec.Provisioner {
+					v.volRepPVCs = append(v.volRepPVCs, *pvc)
+					replicationClassMatchFound = true
+
+					break
+				}
+			}
+
+			if !replicationClassMatchFound {
+				v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+			}
+		default:
+			v.log.Info("filtering using peerClasses")
+
+			replicationClass, err := v.findReplicationClassUsingPeerClasses(peerClasses, storageClass, pvc)
+			if err != nil {
+				return err
+			}
+
+			if replicationClass == nil && !v.instance.Spec.VolSync.Disabled {
+				// ignoring volsnapClass till we store it in the future
+				if _, err := v.findVolSnapClass(storageClass, pvc); err != nil {
+					return err
+				}
+			}
 		}
 	}
 
 	v.log.Info(fmt.Sprintf("Found %d PVCs targeted for VolRep and %d targeted for VolSync",
 		len(v.volRepPVCs), len(v.volSyncPVCs)))
 
+	if !(len(pvcList.Items) == (len(v.volRepPVCs) + len(v.volSyncPVCs))) {
+		return fmt.Errorf("no PVCs are procted")
+	}
+
+	return nil
+}
+
+func (v *VRGInstance) findReplicationClassUsingPeerClasses(peerClasses []ramendrv1alpha1.PeerClass,
+	storageClass *storagev1.StorageClass, pvc *corev1.PersistentVolumeClaim) (
+	*volrep.VolumeReplicationClass, error,
+) {
+	scName := pvc.Spec.StorageClassName
+
+	// this check is needed because we use storageID to compare with other VRC or VSC
+	if _, ok := storageClass.Labels[StorageIDLabel]; !ok {
+		msg := fmt.Sprintf("storageID label not found in storageClass for PVC %s", pvc.Name)
+		v.updatePVCDataReadyCondition(pvc.Namespace, pvc.Name, VRGConditionReasonStorageIDNotFound, msg)
+
+		return nil, fmt.Errorf(msg)
+	}
+
+	peerClass := v.findPeerClassMatchingSC(peerClasses, *scName)
+	if peerClass == nil {
+		msg := fmt.Sprintf("peerClass matching storageClass %s not found for async PVC", *scName)
+		v.updatePVCDataReadyCondition(pvc.Namespace, pvc.Name, VRGConditionReasonPeerClassNotFound, msg)
+
+		return nil, fmt.Errorf(msg)
+	}
+
+	replicationClass := v.findMatchingReplicationClass(peerClass, storageClass, pvc)
+
+	return replicationClass, nil
+}
+
+func (v *VRGInstance) findVolSnapClass(storageClass *storagev1.StorageClass,
+	pvc *corev1.PersistentVolumeClaim,
+) (*snapv1.VolumeSnapshotClass, error) {
+	v.log.Info(fmt.Sprintf("finding snapshotClass for PVC %s with storageClassName %s", pvc.Name, storageClass.Name))
+
+	SnaphotClasses, err := v.volSyncHandler.GetVolumeSnapshotClasses()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, snapshotClass := range SnaphotClasses {
+		if v.selectVSC(snapshotClass, *storageClass) {
+			v.log.Info(fmt.Sprintf("snapshotClass Found: %s", snapshotClass.Name))
+			v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+
+			return &snapshotClass, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find VolumeSnapshotClass")
+}
+
+func (v *VRGInstance) selectVSC(snapshotClass snapv1.VolumeSnapshotClass,
```

Comment:
The current function name is misleading for a boolean return type. Consider renaming it to something such as isVSCSelectable or a similar descriptive name.